iree/turbine/aot/build_actions.py

```diff
@@ -0,0 +1,206 @@
+# Copyright 2024 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from typing import Callable
+
+from abc import abstractmethod, ABC
+from pathlib import Path
+import typing
+import types
+
+import inspect
+
+from iree.build.executor import ActionConcurrency, BuildAction, BuildContext, BuildFile
+from iree.turbine.aot.exporter import ExportOutput
+
+__all__ = [
+    "turbine_generate",
+]
+
+
+def turbine_generate(
+    generator: Callable,
+    *args,
+    name: str,
+    out_of_process: bool = True,
+    **kwargs,
+):
+    """Invokes a user-defined generator callable as an action, performing turbine
+    import and storing the resulting artifacts as outputs.
+
+    Because torch-based generation is usually quite slow and a bottleneck, this
+    action takes pains to use the out of process action pool, allowing multiple
+    generation activities to take place concurrently. Since this requires interacting
+    with the pickle infrastructure, it puts some constraints on usage:
+
+    * generator must be a pickleable callable. In practice, this means that it must
+      be a named function at module scope (without decorator) or a named class at
+      module scope with a `__call__` method.
+    * args and kwargs must be pickleable. In practice, this means primitive values.
+
+    Arguments to the generator are taken from the positional and unmatched keyword
+    arguments passed to `turbine_generate`.
+
+    The generator makes artifacts available as outputs by returning corresponding
+    Python instances (which must be declared as typing parameters for the remoting
+    to work):
+
+    * `ExportOutput`: The result of calling `aot.export(...)` will result in
+      `save_mlir()` being called on it while still in the subprocess to write to
+      a file names `{name}.mlir` if there is one return or `{name}_{n}.mlir` if
+      multiple.
```

Comment:
Can this output `.mlirbc`? What about `.irpa` files?

Reply:
It needs to grow some options for mlirbc. Have some thoughts on how to plug parameter generation through in an ergonomic way. One of:

* Support another return type and then the generator just returns something that causes the export.
* Have an `extra_outputs=` dict that lets you set up additional output files and have that passed into the generator function to do with as you please.

Reply:
(but wanted to land this before adding more)